Lets import all important lidraries to perform data cleaning and aslo the data set that was downloaded from kaggel 


In [10]:
import pandas as pd 
import numpy as np 
df=pd.read_csv('netflix_titles.csv')
df.head()

show_id     type                  title         director  \
0      s1    Movie   Dick Johnson Is Dead  Kirsten Johnson   
1      s2  TV Show          Blood & Water              NaN   
2      s3  TV Show              Ganglands  Julien Leclercq   
3      s4  TV Show  Jailbirds New Orleans              NaN   
4      s5  TV Show           Kota Factory              NaN   

                                                cast        country  \
0                                                NaN  United States   
1  Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   South Africa   
2  Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...            NaN   
3                                                NaN            NaN   
4  Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...          India   

           date_added  release_year rating   duration  \
0  September 25, 2021          2020  PG-13     90 min   
1  September 24, 2021          2021  TV-MA  2 Seasons   
2  September 24, 2021          2021  TV-MA   1 Season   
3  September 24, 2021          2021  TV-MA   1 Season   
4  September 24, 2021          2021  TV-MA  2 Seasons   

                                           listed_in  \
0                                      Documentaries   
1    International TV Shows, TV Dramas, TV Mysteries   
2  Crime TV Shows, International TV Shows, TV Act...   
3                             Docuseries, Reality TV   
4  International TV Shows, Romantic TV Shows, TV ...   

                                         description  
0  As her father nears the end of his life, filmm...  
1  After crossing paths at a party, a Cape Town t...  
2  To protect his family from a powerful drug lor...  
3  Feuds, flirtations and toilet talk go down amo...  
4  In a city of coaching centers known to train I...

Lets see the size of the dataset 

In [8]:
print(df.shape)

(8807, 12)


So there are total of 8807 rows and 12 columns

Lets Find out total null values for each column

In [9]:

df.isnull().sum()

show_id            0
type               0
title              0
director        2634
cast             825
country          831
date_added        10
release_year       0
rating             4
duration           3
listed_in          0
description        0
dtype: int64

We cant delete the rows with null values in director as it make us to delete huge ammount of data insteaded fill it with unknown same with cast,country


In [12]:
df.fillna({
    'director': 'Unknown',
    'cast': 'Unknown',
    'country': 'Unknown'
}, inplace=True)


We has 10 null values in date_added but null values in release_year so we can just fill the null values with release year


In [14]:
df.fillna({
    'date_added':'January 1, ' + df['release_year'].astype(str)},
    inplace=True)


Findout the unique values in rating

In [17]:
print(df['rating'].unique())
print(df['rating'].mode())


['PG-13' 'TV-MA' 'PG' 'TV-14' 'TV-PG' 'TV-Y' 'TV-Y7' 'R' 'TV-G' 'G'
 'NC-17' '74 min' '84 min' '66 min' 'NR' nan 'TV-Y7-FV' 'UR']
0    TV-MA
Name: rating, dtype: object


As 'TV-MA' is mode for rating fill null values with it

In [18]:
df.fillna({
    'rating':'TV-MA'
}, inplace=True)


Now do the same thing for duration

In [19]:
print(df['duration'].unique())
print(df['duration'].mode())


['90 min' '2 Seasons' '1 Season' '91 min' '125 min' '9 Seasons' '104 min'
 '127 min' '4 Seasons' '67 min' '94 min' '5 Seasons' '161 min' '61 min'
 '166 min' '147 min' '103 min' '97 min' '106 min' '111 min' '3 Seasons'
 '110 min' '105 min' '96 min' '124 min' '116 min' '98 min' '23 min'
 '115 min' '122 min' '99 min' '88 min' '100 min' '6 Seasons' '102 min'
 '93 min' '95 min' '85 min' '83 min' '113 min' '13 min' '182 min' '48 min'
 '145 min' '87 min' '92 min' '80 min' '117 min' '128 min' '119 min'
 '143 min' '114 min' '118 min' '108 min' '63 min' '121 min' '142 min'
 '154 min' '120 min' '82 min' '109 min' '101 min' '86 min' '229 min'
 '76 min' '89 min' '156 min' '112 min' '107 min' '129 min' '135 min'
 '136 min' '165 min' '150 min' '133 min' '70 min' '84 min' '140 min'
 '78 min' '7 Seasons' '64 min' '59 min' '139 min' '69 min' '148 min'
 '189 min' '141 min' '130 min' '138 min' '81 min' '132 min' '10 Seasons'
 '123 min' '65 min' '68 min' '66 min' '62 min' '74 min' '131 min' '39 min'
 '46 m

AS it has large number of unique values it cant be filled with mode so just drop null values

In [21]:
df['duration']= df['duration'].dropna()

Now lets findout the duplicates in the dataset

In [22]:
# lers find total number of duplicates in the dataset
df.duplicated().sum()

0

AS there are no duplicates in the data set no need to perform anything

Lets find out the data type of each column 

In [23]:

df.dtypes

show_id         object
type            object
title           object
director        object
cast            object
country         object
date_added      object
release_year     int64
rating          object
duration        object
listed_in       object
description     object
dtype: object

Lets convert the data type of date_added to Date time


In [24]:
df['date_added'].unique()

array(['September 25, 2021', 'September 24, 2021', 'September 23, 2021',
       ..., 'December 6, 2018', 'March 9, 2016', 'January 11, 2020'],
      dtype=object)

In [26]:
df['date_added'] = df['date_added'].str.strip()  # remove leading/trailing spaces
df['date_added'] = pd.to_datetime(df['date_added'], format="%B %d, %Y")


In [27]:
print(df['country'].unique())

['United States' 'South Africa' 'Unknown' 'India'
 'United States, Ghana, Burkina Faso, United Kingdom, Germany, Ethiopia'
 'United Kingdom' 'Germany, Czech Republic' 'Mexico' 'Turkey' 'Australia'
 'United States, India, France' 'Finland' 'China, Canada, United States'
 'South Africa, United States, Japan' 'Nigeria' 'Japan'
 'Spain, United States' 'France' 'Belgium' 'United Kingdom, United States'
 'United States, United Kingdom' 'France, United States' 'South Korea'
 'Spain' 'United States, Singapore' 'United Kingdom, Australia, France'
 'United Kingdom, Australia, France, United States'
 'United States, Canada' 'Germany, United States'
 'South Africa, United States' 'United States, Mexico'
 'United States, Italy, France, Japan'
 'United States, Italy, Romania, United Kingdom'
 'Australia, United States' 'Argentina, Venezuela'
 'United States, United Kingdom, Canada' 'China, Hong Kong' 'Russia'
 'Canada' 'Hong Kong' 'United States, China, Hong Kong'
 'Italy, United States' 'United Sta

Interview questions:


1.What are missing values and how do you handle them?

ans:Missing values are gaps or empty spots in your data where some information is missing. For example, if a movie’s 
director name isn’t recorded, that cell is a missing value.

It can be fixed using :

Fill them in with a placeholder like "Unknown"  which I used for director column

Remove rows or columns that have too many missing values which I used for duration column

Estimate or predict the missing values based on other data which I used for Date_added  and rating column


2.How do you treat duplicate records?

ans:

Duplicate records can be removed from the data maintaing only one

3.Difference between dropna() and fillna() in Pandas?

ans:

dropna() — removes rows or columns that contain missing values (NaNs)

fillna() — fills the missing values with a specific value you choose


4.What is outlier treatment and why is it important?

ans:

Outlier treatment means identifying and handling data points that are very different from the rest of your data

Outliers can distort your analysis and lead to wrong conclusions.


5.Explain the process of standardizing data.

Standardizing data means transforming your data so it has a mean (average) of 0 and a standard deviation of 1. This makes different features comparable and helps many algorithms work better.


6.How do you handle inconsistent data formats (e.g., date/time)?

Identify the inconsistencies:
Check your data to find different formats (e.g., "2023-06-01", "06/01/2023", "June 1, 2023").

Choose a standard format:
Decide on one consistent format for the data, like YYYY-MM-DD.

Convert all data to the standard format:
Use tools or code to parse and convert the different formats into your chosen format.

7.What are common data cleaning challenges?

Here are some common data cleaning challenges explained simply:

Missing Data:
Some values are missing or blank, which can mess up analysis.

Duplicate Records:
Same data repeated multiple times, leading to biased results.

Inconsistent Formats:
Data like dates, phone numbers, or text might have different formats.

Incorrect Data:
Errors or typos in the data, like a wrong age or misspelled names.

Outliers:
Extreme values that don’t fit the pattern and can skew results.

Noise:
Irrelevant or random errors in the data.

Data Integration Issues:
Combining data from multiple sources can cause mismatches or conflicts.

Unstructured Data:
Data in messy formats like free text, making it hard to analyze.

8.How can you check data quality?

To check data quality, you can:

Look for Missing Values:
Use functions like isnull() or info() in Pandas to see if data is missing.

Check for Duplicates:
Use duplicated() to find repeated rows.

Validate Data Types:
Ensure columns have the correct data types (e.g., dates are dates, numbers are numbers).

Examine Summary Statistics:
Use describe() to spot unusual values or outliers.

Check for Consistency:
Verify if categorical data uses consistent labels (e.g., “Yes” vs “yes”).

Visualize Data:
Plot histograms, boxplots, or scatterplots to identify anomalies or patterns.

Use Data Validation Rules:
Apply rules like “age should be between 0 and 120” to spot errors.

Cross-Check with External Sources:
Compare data against trusted references if available.